### DISCUSSION TITLE
Share your ideas and wishes !

### DISCUSSION THREAD

**Discussion by genaray** [genaray/Arch#8] [Ideas]
So since there actually a few people using this ecs framework, its probably a good idea to respond to their wishes :) 

A few features are currently already planned and being worked on : 
https://github.com/users/genaray/projects/1
- Structural changes
- CommandBuffer to utilize structural changes when multithreading
- EntityTemplates for cloning entities and their values in masses

What else would you love to see ? :) Let me know !

**Comment by MindSwipe** [comment]
A collection of things I'd love to see:

- Changing an Entity's Archetype at runtime (probably best to delay until a CommandBuffer has been implemented)

- Exclusive Queries
A way to query for Entities that contain an exact match of components, so say we execute a Query with `Exclusive = { Transform, Roation }` in a world with an Entity (A) with components Transform and Rotation and another Entity (B) with Transform, Rotation and Ai, only Entity A would be returned

- Registering Systems in a World
A built in way of registering/ grouping systems would be great, so we can setup a world, register systems, and then every frame/ simulation step just call `world.Run`. This would mean we'd also need some way of ordering/ grouping systems, and for that I'd closely mimic what Bevy did, ordered SystemStages that actually "hold" the systems with a way to register SystemStages and set ordering

- System Resources
A way for the World to have state outside of entities would be great, since some things only make sense to have one of, for example Time

P.S maybe the Projects aren't public in this Repo? I get a 404 when I click on the project link.

**Reply by genaray** [reply]
Noted and thanks for your input, bug discoveries and bug fixes ! :) 
And you were right, the projects werent public, just updated that :D 

Im currently working on the structural changes, nearly finished ^^ I will push them soon to a feature branch. 

I love the ideas of exclusive queries, that would be extremely usefull and would also make some tag components obsolete. 
Systems would probably also smaller the boilerplate code which is always welcome :D 

However i think that Systems and World State would be perfect as some sort of Addon/Extensions. They could get their own repo and hook themself into this one for extending its capabilities ^^ This would keep the main code base clean and bare minimum, while the other ones could focus on building a whole "world" around it :)

**Reply by MindSwipe** [reply]
Cool, I can now view the projects, it worked 👍

Ah, so that's what you meant by "structural changes", very cool

I can get to work on exclusive queries if you want me to (with a new issue to group discussions about the API)

Outsourcing Systems and World State sounds like a good plan, so people that don't want/ need those features don't pay for any potential overhead, this is also something I'd be willing to do some work on

**Reply by genaray** [reply]
That sounds great :) 

Of course, i would love some helping hands ! If you want you can also work on both :D 
Alright, should i create the issue or will you do this ? ^^

**Reply by MindSwipe** [reply]
I can do it 👍 

Also, just a comment on the "Refactor Code to full c# code convention" item on the roadmap, the easiest way would probably be to add the StyleCop.Analyzers to all Arch.* projects and "fix" the warnings (which are about 550+. so maybe keep it on the backburner for an "Arch v2.0" release?)

**Reply by genaray** [reply]
Totally, the warnings are extreme... i actually "adjusted" the code style a few days back, however this only partially removed the warnings :D So this would also require some more work... i did the mistake and started this project with some sort of java like syntax since my rider config fucked up somewhere ^^

**Reply by MindSwipe** [reply]
Yup, but a lot of them are asinine, StyleCop by default doesn't like it if you omit the braces for a single line if statement (something [even MS themselves do](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/mono/mono/tests/jit-int.cs#L14)), so coming up with a "sane" ruleset would be first priority when adopting StyleCop

**Reply by genaray** [reply]
Well thats true, i will take care of this soon ^^ 
However, i just finished the structural changes API... entities can now move between archetypes by adding and removing components from them :) 

```csharp
entity.Add<T>(new T());
entity.Add<T>();
entity.Remove<T>();
```

**Reply by genaray** [reply]
Just tested the exclusive queries, nice work ! :) 

Besides that, the command buffers were introduced with #24 ^^ Now its probably time to finally look at extensions like a Systems API and co :)
